In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
%cd drive/MyDrive/Colab Notebooks/adult

/content/drive/MyDrive/Colab Notebooks/adult


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
df_train = pd.read_csv('adult.data', index_col = False, sep=',', names = ['age','workclass','fnlwght', 'education', 'education_num','marital_status','occupation','relationship','race','sex','capital-gain','capital-loss','hours-per-week','native-country','income'])

In [ ]:
df_train.head()

age          workclass  fnlwght   education  education_num  \
0   39          State-gov    77516   Bachelors             13   
1   50   Self-emp-not-inc    83311   Bachelors             13   
2   38            Private   215646     HS-grad              9   
3   53            Private   234721        11th              7   
4   28            Private   338409   Bachelors             13   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [ ]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwght         32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [ ]:
df_train['occupation'].unique()

array([' Adm-clerical', ' Exec-managerial', ' Handlers-cleaners',
       ' Prof-specialty', ' Other-service', ' Sales', ' Craft-repair',
       ' Transport-moving', ' Farming-fishing', ' Machine-op-inspct',
       ' Tech-support', ' ?', ' Protective-serv', ' Armed-Forces',
       ' Priv-house-serv'], dtype=object)

In [ ]:
df_train[df_train['occupation'] == ' ?']

age workclass  fnlwght      education  education_num  \
27      54         ?   180211   Some-college             10   
61      32         ?   293936        7th-8th              4   
69      25         ?   200681   Some-college             10   
77      67         ?   212759           10th              6   
106     17         ?   304873           10th              6   
...    ...       ...      ...            ...            ...   
32530   35         ?   320084      Bachelors             13   
32531   30         ?    33811      Bachelors             13   
32539   71         ?   287372      Doctorate             16   
32541   41         ?   202822        HS-grad              9   
32542   72         ?   129912        HS-grad              9   

               marital_status occupation    relationship                 race  \
27         Married-civ-spouse          ?         Husband   Asian-Pac-Islander   
61      Married-spouse-absent          ?   Not-in-family                White   
69              Never-married          ?       Own-child                White   
77         Married-civ-spouse          ?         Husband                White   
106             Never-married          ?       Own-child                White   
...                       ...        ...             ...                  ...   
32530      Married-civ-spouse          ?            Wife                White   
32531           Never-married          ?   Not-in-family   Asian-Pac-Islander   
32539      Married-civ-spouse          ?         Husband                White   
32541               Separated          ?   Not-in-family                Black   
32542      Married-civ-spouse          ?         Husband                White   

           sex  capital-gain  capital-loss  hours-per-week  native-country  \
27        Male             0             0              60           South   
61        Male             0             0              40               ?   
69        Male             0             0              40   United-States   
77        Male             0             0               2   United-States   
106     Female         34095             0              32   United-States   
...        ...           ...           ...             ...             ...   
32530   Female             0             0              55   United-States   
32531   Female             0             0              99   United-States   
32539     Male             0             0              10   United-States   
32541   Female             0             0              32   United-States   
32542     Male             0             0              25   United-States   

       income  
27       >50K  
61      <=50K  
69      <=50K  
77      <=50K  
106     <=50K  
...       ...  
32530    >50K  
32531   <=50K  
32539    >50K  
32541   <=50K  
32542   <=50K  

[1843 rows x 15 columns]

In [ ]:
sum(df_train['occupation'] == ' ?')

1843

In [ ]:
sum(df_train['workclass'] == ' ?')

1836

In [ ]:
sum(df_train['native-country'] == ' ?')

583

In [ ]:
## remove row with ?
df_new = df_train[df_train['occupation'] != ' ?']
df_new = df_new[df_new['workclass'] != ' ?']
df_new = df_new[df_new['native-country'] != ' ?']

In [ ]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30162 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             30162 non-null  int64 
 1   workclass       30162 non-null  object
 2   fnlwght         30162 non-null  int64 
 3   education       30162 non-null  object
 4   education_num   30162 non-null  int64 
 5   marital_status  30162 non-null  object
 6   occupation      30162 non-null  object
 7   relationship    30162 non-null  object
 8   race            30162 non-null  object
 9   sex             30162 non-null  object
 10  capital-gain    30162 non-null  int64 
 11  capital-loss    30162 non-null  int64 
 12  hours-per-week  30162 non-null  int64 
 13  native-country  30162 non-null  object
 14  income          30162 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [ ]:
# remove fnlweight
df_new = df_new.drop(['fnlwght'],axis=1)

In [ ]:
df_new

age          workclass    education  education_num  \
0       39          State-gov    Bachelors             13   
1       50   Self-emp-not-inc    Bachelors             13   
2       38            Private      HS-grad              9   
3       53            Private         11th              7   
4       28            Private    Bachelors             13   
...    ...                ...          ...            ...   
32556   27            Private   Assoc-acdm             12   
32557   40            Private      HS-grad              9   
32558   58            Private      HS-grad              9   
32559   22            Private      HS-grad              9   
32560   52       Self-emp-inc      HS-grad              9   

            marital_status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital-gain  capital-loss  hours-per-week  native-country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[30162 rows x 14 columns]

In [ ]:
# convert categorical sex column to numerical
df_new['sex'].replace(' Male', 0, inplace=True)
df_new['sex'].replace(' Female', 1, inplace=True)
# df_new = df_new['sex'].replace(' Female', 1)

In [ ]:
df_new

age          workclass    education  education_num  \
0       39          State-gov    Bachelors             13   
1       50   Self-emp-not-inc    Bachelors             13   
2       38            Private      HS-grad              9   
3       53            Private         11th              7   
4       28            Private    Bachelors             13   
...    ...                ...          ...            ...   
32556   27            Private   Assoc-acdm             12   
32557   40            Private      HS-grad              9   
32558   58            Private      HS-grad              9   
32559   22            Private      HS-grad              9   
32560   52       Self-emp-inc      HS-grad              9   

            marital_status          occupation    relationship    race  sex  \
0            Never-married        Adm-clerical   Not-in-family   White    0   
1       Married-civ-spouse     Exec-managerial         Husband   White    0   
2                 Divorced   Handlers-cleaners   Not-in-family   White    0   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black    0   
4       Married-civ-spouse      Prof-specialty            Wife   Black    1   
...                    ...                 ...             ...     ...  ...   
32556   Married-civ-spouse        Tech-support            Wife   White    1   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White    0   
32558              Widowed        Adm-clerical       Unmarried   White    1   
32559        Never-married        Adm-clerical       Own-child   White    0   
32560   Married-civ-spouse     Exec-managerial            Wife   White    1   

       capital-gain  capital-loss  hours-per-week  native-country  income  
0              2174             0              40   United-States   <=50K  
1                 0             0              13   United-States   <=50K  
2                 0             0              40   United-States   <=50K  
3                 0             0              40   United-States   <=50K  
4                 0             0              40            Cuba   <=50K  
...             ...           ...             ...             ...     ...  
32556             0             0              38   United-States   <=50K  
32557             0             0              40   United-States    >50K  
32558             0             0              40   United-States   <=50K  
32559             0             0              20   United-States   <=50K  
32560         15024             0              40   United-States    >50K  

[30162 rows x 14 columns]

In [ ]:
df_new['race'].unique()

array([' White', ' Black', ' Asian-Pac-Islander', ' Amer-Indian-Eskimo',
       ' Other'], dtype=object)

In [ ]:
from sklearn.preprocessing import LabelEncoder

# convert income to 0,1 (0: <=50K, 1: >=50K)
le = LabelEncoder()
le.fit(df_new['income'])
df_new['income'] = le.transform(df_new['income'])

In [ ]:
df_new

age          workclass    education  education_num  \
0       39          State-gov    Bachelors             13   
1       50   Self-emp-not-inc    Bachelors             13   
2       38            Private      HS-grad              9   
3       53            Private         11th              7   
4       28            Private    Bachelors             13   
...    ...                ...          ...            ...   
32556   27            Private   Assoc-acdm             12   
32557   40            Private      HS-grad              9   
32558   58            Private      HS-grad              9   
32559   22            Private      HS-grad              9   
32560   52       Self-emp-inc      HS-grad              9   

            marital_status          occupation    relationship    race  sex  \
0            Never-married        Adm-clerical   Not-in-family   White    0   
1       Married-civ-spouse     Exec-managerial         Husband   White    0   
2                 Divorced   Handlers-cleaners   Not-in-family   White    0   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black    0   
4       Married-civ-spouse      Prof-specialty            Wife   Black    1   
...                    ...                 ...             ...     ...  ...   
32556   Married-civ-spouse        Tech-support            Wife   White    1   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White    0   
32558              Widowed        Adm-clerical       Unmarried   White    1   
32559        Never-married        Adm-clerical       Own-child   White    0   
32560   Married-civ-spouse     Exec-managerial            Wife   White    1   

       capital-gain  capital-loss  hours-per-week  native-country  income  
0              2174             0              40   United-States       0  
1                 0             0              13   United-States       0  
2                 0             0              40   United-States       0  
3                 0             0              40   United-States       0  
4                 0             0              40            Cuba       0  
...             ...           ...             ...             ...     ...  
32556             0             0              38   United-States       0  
32557             0             0              40   United-States       1  
32558             0             0              40   United-States       0  
32559             0             0              20   United-States       0  
32560         15024             0              40   United-States       1  

[30162 rows x 14 columns]

In [ ]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30162 entries, 0 to 32560
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             30162 non-null  int64 
 1   workclass       30162 non-null  object
 2   education       30162 non-null  object
 3   education_num   30162 non-null  int64 
 4   marital_status  30162 non-null  object
 5   occupation      30162 non-null  object
 6   relationship    30162 non-null  object
 7   race            30162 non-null  object
 8   sex             30162 non-null  int64 
 9   capital-gain    30162 non-null  int64 
 10  capital-loss    30162 non-null  int64 
 11  hours-per-week  30162 non-null  int64 
 12  native-country  30162 non-null  object
 13  income          30162 non-null  int64 
dtypes: int64(7), object(7)
memory usage: 3.5+ MB


In [ ]:
# create dummy values for other categorical column
df_dummy = pd.get_dummies(df_new, drop_first=True)

In [ ]:
df_dummy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30162 entries, 0 to 32560
Data columns (total 96 columns):
 #   Column                                      Non-Null Count  Dtype
---  ------                                      --------------  -----
 0   age                                         30162 non-null  int64
 1   education_num                               30162 non-null  int64
 2   sex                                         30162 non-null  int64
 3   capital-gain                                30162 non-null  int64
 4   capital-loss                                30162 non-null  int64
 5   hours-per-week                              30162 non-null  int64
 6   income                                      30162 non-null  int64
 7   workclass_ Local-gov                        30162 non-null  uint8
 8   workclass_ Private                          30162 non-null  uint8
 9   workclass_ Self-emp-inc                     30162 non-null  uint8
 10  workclass_ Self-emp-not-inc       

In [ ]:
X = df_dummy.drop(['income'],axis=1)
y = df_dummy['income']

In [ ]:
from sklearn.model_selection import train_test_split, cross_val_score

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 101)

In [ ]:
# 1st method: Decision tree
from sklearn.tree import DecisionTreeClassifier

# use cross-validation to get the accuracy
dt = DecisionTreeClassifier()
cv = cross_val_score(dt, X_train, y_train, cv=5)
cv.mean(), cv.std()

(0.8126734677663519, 0.003174366996683013)

In [ ]:
from sklearn.metrics import confusion_matrix, classification_report
dt.fit(X_train, y_train)
y_pred = dt.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[3935  541]
 [ 613  944]]
              precision    recall  f1-score   support

           0       0.87      0.88      0.87      4476
           1       0.64      0.61      0.62      1557

    accuracy                           0.81      6033
   macro avg       0.75      0.74      0.75      6033
weighted avg       0.81      0.81      0.81      6033



In [ ]:
X = df_dummy.drop(['income'],axis=1)
y = df_dummy['income']

In [ ]:
# Feature selection: remove some unwanted columns using pearson correlation
import seaborn as sns
cor = df_dummy.corr()

In [ ]:
#Correlation with output variable (income)
cor_target = abs(cor["income"])
#Selecting highly correlated features
# cor_target[cor_target>0.1]
relevant_features = cor_target[cor_target>0.1]

In [ ]:
relevant_features = relevant_features.index.to_list()

In [ ]:
len(relevant_features)

23

In [ ]:
df_feature_selected = df_dummy[relevant_features]

In [ ]:
X_new = df_feature_selected.drop(['income'], axis=1)
y_new = df_feature_selected['income']

In [ ]:
X_train_new, X_test_new, y_train_new, y_test_new = train_test_split(X_new, y_new, test_size = 0.2, random_state = 101)

In [ ]:
cv = cross_val_score(dt, X_new, y_new, cv=5)
cv.mean(), cv.std()

(0.8174858184934234, 0.006502149876168235)

In [ ]:
dt.fit(X_train_new, y_train_new)
y_pred = dt.predict(X_test_new)
print(classification_report(y_test_new, y_pred))

              precision    recall  f1-score   support

           0       0.86      0.91      0.88      4476
           1       0.68      0.58      0.63      1557

    accuracy                           0.82      6033
   macro avg       0.77      0.74      0.76      6033
weighted avg       0.82      0.82      0.82      6033



In [ ]:
# hyperparameter-tuning 
# before feature selection
from sklearn.model_selection import GridSearchCV

parameters = {'criterion':['gini','entropy'],
              'min_samples_split': [4,5,10],
              'max_depth':[3,6,9]}

dt = DecisionTreeClassifier()
gridcv = GridSearchCV(dt, parameters)
gridcv.fit(X_train, y_train)


GridSearchCV(estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [3, 6, 9],
                         'min_samples_split': [4, 5, 10]})

In [ ]:
gridcv.best_params_

{'criterion': 'entropy', 'max_depth': 9, 'min_samples_split': 5}

In [ ]:
y_pred = gridcv.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      0.96      0.90      4476
           1       0.82      0.54      0.65      1557

    accuracy                           0.85      6033
   macro avg       0.84      0.75      0.78      6033
weighted avg       0.85      0.85      0.84      6033



In [ ]:
# gridcv after feature selection
parameters = {'criterion':['gini','entropy'],
              'min_samples_split': [4,5,10],
              'max_depth':[3,6,9]}

dt = DecisionTreeClassifier()
gridcv = GridSearchCV(dt, parameters)
gridcv.fit(X_train_new, y_train_new)

print(gridcv.best_params_)

y_pred = gridcv.predict(X_test_new)
print(classification_report(y_test_new, y_pred))

{'criterion': 'gini', 'max_depth': 9, 'min_samples_split': 5}
              precision    recall  f1-score   support

           0       0.87      0.95      0.91      4476
           1       0.80      0.58      0.67      1557

    accuracy                           0.85      6033
   macro avg       0.83      0.76      0.79      6033
weighted avg       0.85      0.85      0.85      6033



before feature selection, accuracy = 81.27% <br>
try feature selection <br>
94 feaures -> 22 features <br>
after feature selection, accuracy = 81.72% <br>
therefore, no significant changes <br>
<br>
try hyperparameter-tuning using GridSearchCV to find the best parameters (criterion, min_samples_split, max_depth) of decision tree
<br>
best params: {'criterion': 'gini', 'max_depth': 9, 'min_samples_split': 4}
<br>
accuracy after hyperparameter-tuning: 85%

In [ ]:
# 2nd method: SVM
from sklearn.svm import SVC

# original dataset
svc = SVC()
svc.fit(X_train, y_train)

y_pred = svc.predict(X_test)

In [ ]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.79      0.97      0.87      4476
           1       0.74      0.27      0.39      1557

    accuracy                           0.79      6033
   macro avg       0.77      0.62      0.63      6033
weighted avg       0.78      0.79      0.75      6033



In [ ]:
# original dataset cross validation
svc = SVC()
cv = cross_val_score(svc, X, y, cv=5)
cv.mean(), cv.std()

(0.7960015779701474, 0.0021206491281095946)

In [ ]:
# feature selected dataset
svc = SVC()
svc.fit(X_train_new, y_train_new)

y_pred = svc.predict(X_test_new)
print(classification_report(y_test_new, y_pred))

              precision    recall  f1-score   support

           0       0.79      0.97      0.87      4476
           1       0.74      0.27      0.39      1557

    accuracy                           0.79      6033
   macro avg       0.77      0.62      0.63      6033
weighted avg       0.78      0.79      0.75      6033



In [ ]:
# feature selected dataset cross validation
svc = SVC()
cv = cross_val_score(svc, X_new, y_new, cv=5)
cv.mean(), cv.std()

(0.7960015779701474, 0.0021206491281095946)

SVM <br>
before feature selection accuracy: 79%
<br>
after feature selection accuracy: 79%
<br>
no significant changes
<br>


In [ ]:
# hyperparameter-tuning
# before feature selection

svc = SVC()
parameters = {'C':[1.0, 5.0],
              'kernel':['rbf','poly']}

gridcv = GridSearchCV(svc, parameters)

gridcv.fit(X_train, y_train)

print(gridcv.best_params_)

y_pred = gridcv.predict(X_test)
print(classification_report(y_test, y_pred))

{'C': 5.0, 'kernel': 'rbf'}
              precision    recall  f1-score   support

           0       0.79      0.97      0.87      4476
           1       0.75      0.27      0.39      1557

    accuracy                           0.79      6033
   macro avg       0.77      0.62      0.63      6033
weighted avg       0.78      0.79      0.75      6033

